numpy/core/numeric.py

```diff
@@ -289,16 +288,14 @@ def full(shape, fill_value, dtype=None, order='C'):
     >>> np.full((2, 2), np.inf)
     array([[ inf,  inf],
            [ inf,  inf]])
-    >>> np.full((2, 2), 10, dtype=np.int)
+    >>> np.full((2, 2), 10)
     array([[10, 10],
            [10, 10]])
 
     """
+    if dtype is None:
+        dtype = array(fill_value).dtype
```

Comment:
Well, `np.full(3, (1, 2))` (with or without passing `dtype=object`) didn't work and still doesn't work.  The real fun happens when someone write `np.full(3, (1, 2, 3))` though (it used to, and still, returns `array([1, 2, 3])`).

Is there a function equivalent to `array(fill_value).dtype`, but that would actually return `object` when a non-scalar is passed in?  I thought `np.obj2sctype(..., default=object)` would work, but it returns the dtype of a ndarray when a ndarray is passed in (so `np.full(3, np.array([1, 2, 3]))` would still "fail").  (By the way I find `obj2sctype's` "return None if everything fails" somewhat unpythonic.)